### PR TITLE
[Silabs] Enabling custom board support for matter

### PR DESCRIPTION
**Pull request by fuxiaoming-lumi**
Enable Custom Board Support:
- This enhancement allows for greater flexibility in selecting mcu models, leveraging the diverse range of mcu models provided by Silabs to cater to various application requirements.
- When integrating the connectedhomeip repo as a third-party submodule into developers' projects, the objective is to ensure a seamless downstream process. This involves leveraging `sl_pre_gen_path` to adapt flexibly to custom board configurations.
 ```
  ├── smart-plug-app
  │   ├── silabs
  │   │   ├── args.gni
  │   │   ├── BUILD.gn
  │   │   ├── slcp
  │   │   │   ├── CUSTOM
  │   │   │   │   ├── autogen
  │   │   │   │   └── config
  │   │   │   ├── gatt_configuration.btconf
  │   │   │   └── matter-platform.slcp
  │   │   ├── src
  │   └── smart-plug-common
  │       ├── BUILD.gn
  │       ├── smart-plug.matter
  │       └── smart-plug.zap
  └── third-party
      ├── connectedhomeip
 ```

Testing done:
```
./scripts/examples/gn_silabs_example.sh examples/lighting-app/silabs out CUSTOM silabs_family=\"efr32mg24\" silabs_mcu=\"EFR32MG24A410F1536IM48\" sl_pre_gen_path=\"<PATH_TO_SLCP_GEN>\"
```

### CODE REVIEW

**Review by kiel-apple — APPROVED** [review]

**Review by arkq — APPROVED** [review]

**Comment by fuxiaoming-lumi** [comment]
@jmartinez-silabs Could you give it a quick review？Thanks.

**Review by jmartinez-silabs — APPROVED** [review]
Thank you @fuxiaoming-lumi that's a good idea

File: third_party/silabs/silabs_board.gni [hunk 1]
```diff
@@ -150,6 +156,14 @@ if (silabs_board == "BRD4338A" || silabs_board == "BRD2605A") {
   # ThunderBoards don't have a LCD,
   show_qr_code = false
   disable_lcd = true
+
+  # Custom Board ----------
+} else if (silabs_board == "CUSTOM") {
+  print("Using custom board configuration")
+  print("silabs_family:", silabs_family)
+  print("silabs_mcu:", silabs_mcu)
+  assert(silabs_family != "", "Must specify silabs_family for custom board")
```
Comment (jmartinez-silabs) [inline third_party/silabs/silabs_board.gni:165]: Nit: We could put the assert at the end of this file.

For any board at this end of the file a family and MCU must be specified

**Review by jmartinez-silabs — APPROVED** [review]

File: third_party/silabs/silabs_board.gni [hunk 2]
```diff
@@ -13,9 +13,15 @@
 # limitations under the License.
 
 declare_args() {
-  # EFR32 board used
+  # Silabs board used
```
Comment (andy31415) [inline third_party/silabs/silabs_board.gni:16]: Would there be a way to add comments that enhance reading the code (for example we could add examples or additional explanation)?

Reading here, it seems obvious that `silabs_board` is likely the `Silabs board` and I believe this applies to family and mcu as well.

File: third_party/silabs/silabs_board.gni (hunk 1 above)
Comment (andy31415) [inline third_party/silabs/silabs_board.gni:165]:
```suggestion
  assert(silabs_family != "", "Must specify silabs_family for a custom board")
```

File: third_party/silabs/silabs_board.gni [hunk 3]
```diff
@@ -150,6 +156,14 @@ if (silabs_board == "BRD4338A" || silabs_board == "BRD2605A") {
...
   show_qr_code = false
   disable_lcd = true
+
+  # Custom Board ----------
+} else if (silabs_board == "CUSTOM") {
+  print("Using custom board configuration")
+  print("silabs_family:", silabs_family)
+  print("silabs_mcu:", silabs_mcu)
+  assert(silabs_family != "", "Must specify silabs_family for custom board")
+  assert(silabs_mcu != "", "Must specify silabs_mcu for custom board")
```
Comment (andy31415) [inline third_party/silabs/silabs_board.gni:166]:
```suggestion
  assert(silabs_mcu != "", "Must specify silabs_mcu for a custom board")
```

**Review by andy31415 — APPROVED** [review]

File: third_party/silabs/silabs_board.gni (hunk 1 above)
Comment (fuxiaoming-lumi) [reply third_party/silabs/silabs_board.gni]: Done

File: third_party/silabs/silabs_board.gni (hunk 2 above)
Comment (fuxiaoming-lumi) [reply third_party/silabs/silabs_board.gni]: Add additional explanation.

File: third_party/silabs/silabs_board.gni (hunk 3 above)
Comment (fuxiaoming-lumi) [reply third_party/silabs/silabs_board.gni]: Put the assert at the end

File: third_party/silabs/silabs_board.gni (hunk 1 above)
Comment (fuxiaoming-lumi) [reply third_party/silabs/silabs_board.gni]: Put the assert at the end

**Pull request merged**